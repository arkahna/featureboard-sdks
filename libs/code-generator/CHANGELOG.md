# @featureboard/code-generator

## 0.3.6

### Patch Changes

- 59c8364: Added namespace validator to cli

## 0.3.5

### Patch Changes

- 69510ca: Fix code gen generator
  Add add target executor

## 0.3.4

### Patch Changes

- 3041a28: Move token management code to new project api-authentication and allow nx-plugin to make use of it.

## 0.3.3

### Patch Changes

- 9f1d891: Better error logging for unauthorized requests

## 0.3.2

### Patch Changes

- c0a2eab: Fix build process for NX plugin

## 0.3.1

### Patch Changes

- cdc1490: Change the display name of project to product

## 0.3.0

### Minor Changes

- 9e22fb1: Add typescript code generator template

## 0.2.1

### Patch Changes

- 5755d8f: Updated documentation and instructions

## 0.2.0

### Minor Changes

- bc4d42a: Add login, select-organization and account commands to CLI

## 0.1.1

### Patch Changes

- 41f499f: Fix package process

## 0.1.0

### Minor Changes

- 7587dae: Add support for modern bundlers

## 0.0.2

### Patch Changes

- [#42](https://github.com/arkahna/featureboard-sdks/pull/42) [`777b2f1`](https://github.com/arkahna/featureboard-sdks/commit/777b2f1556f718d3e5799878f83a8519e855cfa1) Thanks [@emlyn-arkahna](https://github.com/emlyn-arkahna)! - Update user messages
