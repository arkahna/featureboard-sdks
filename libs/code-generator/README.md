# FeatureBoard Code Generator

NX tool that generates FeatureBoard SDKs for different languages.

![npm](https://img.shields.io/npm/v/%40featureboard%2Fcode-generator?logo=npm) ![npm](https://img.shields.io/npm/dt/%40featureboard%2Fcode-generator) ![GitHub Workflow](https://img.shields.io/github/actions/workflow/status/arkahna/featureboard-sdks/main.yml?logo=github)

## Installation

```bash
npm add @featureboard/code-generator
```
```bash
yarn add @featureboard/code-generator
```
```bash
pnpm add @featureboard/code-generator
```

## Release Notes

Our changelog can be found on [GitHub](https://github.com/arkahna/featureboard-sdks/blob/main/libs/code-generator/CHANGELOG.md).
