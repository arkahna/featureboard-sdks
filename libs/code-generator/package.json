{
    "name": "@featureboard/code-generator",
    "version": "0.3.0",
    "main": "src/index.js",
    "license": "MIT",
    "sideEffects": false,
    "repository": {
        "type": "git",
        "url": "git+https://github.com/arkahna/featureboard-sdks.git"
    },
    "keywords": [
        "Feature management",
        "Feature toggles",
        "Feature flags"
    ],
    "bugs": {
        "url": "https://github.com/arkahna/featureboard-sdks/issues"
    },
    "homepage": "https://github.com/arkahna/featureboard-sdks/tree/master/libs/code-generator#readme"
}
