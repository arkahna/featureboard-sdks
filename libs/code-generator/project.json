{
    "name": "code-generator",
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "sourceRoot": "libs/code-generator/src",
    "projectType": "library",
    "targets": {
        "build": {
            "executor": "@nx/js:tsc",
            "outputs": ["{options.outputPath}"],
            "options": {
                "outputPath": "dist/libs/code-generator",
                "tsConfig": "libs/code-generator/tsconfig.lib.json",
                "packageJson": "libs/code-generator/package.json",
                "main": "libs/code-generator/src/index.ts",
                "assets": [
                    "libs/code-generator/*.md",
                    {
                        "input": "./libs/code-generator/src",
                        "glob": "**/*.template",
                        "output": "./src"
                    }
                ]
            }
        },
        "lint": {
            "executor": "@nx/eslint:lint",
            "outputs": ["{options.outputFile}"],
            "options": {
                "lintFilePatterns": [
                    "libs/code-generator/**/*.ts",
                    "libs/code-generator/package.json"
                ]
            }
        }
    },
    "tags": []
}
