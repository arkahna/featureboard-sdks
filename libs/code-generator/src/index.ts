export { getUserOrganizations } from './lib/api/get-organizations'
export { getProjectFeatures } from './lib/api/get-project-features'
export type { FeatureBoardFeature } from './lib/api/get-project-features'
export { getProjects } from './lib/api/get-projects'
export { codeGenerator } from './lib/code-generator'
export type { CodeGeneratorOptions, Template } from './lib/code-generator'
export { queryFeatureBoard } from './lib/queryFeatureBoard'
export type { FeatureBoardAuth } from './lib/queryFeatureBoard'
export { FsTree, flushChanges, printChanges } from './lib/tree/tree'
export type { Tree } from './lib/tree/tree'
