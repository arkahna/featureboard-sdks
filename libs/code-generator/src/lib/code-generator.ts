import path from 'node:path'
import prompts from 'prompts'
import type {
    FeatureBoardFeature,
    FeatureBoardProject,
} from './api/get-project-features'
import { getProjectFeatures } from './api/get-project-features'
import { getProjects } from './api/get-projects'
import {
    getDotNetNameSpace,
    toDotNetType,
    toPascalCase,
} from './generators/dotnet-api/functions'
import { generateFiles } from './generators/dotnet-api/generate-files'
import type { FeatureBoardAuth } from './queryFeatureBoard'
import type { Tree } from './tree/tree'

export type Template = 'dotnet-api' | 'typescript'

export interface CodeGeneratorOptions {
    template: Template
    auth: FeatureBoardAuth
    apiEndpoint?: string
    featureBoardProjectName?: string

    interactive: boolean

    tree: Tree
    relativeFilePath: string
}

export async function codeGenerator(
    options: CodeGeneratorOptions,
): Promise<void> {
    const [project, features] = await getFeatures(options)
    if (!project || !features) {
        return
    }
    console.log(project)
    const relativeFilePath = options.relativeFilePath

    const templateSpecificOptions: any = {}
    switch (options.template) {
        case 'dotnet-api':
            templateSpecificOptions.namespace = getDotNetNameSpace(
                options.tree,
                relativeFilePath,
            )
    }

    await generateFiles(
        options.tree,
        path.join(__dirname, 'templates', options.template),
        options.relativeFilePath,
        {
            toPascalCase,
            toDotNetType,
            ...templateSpecificOptions,
            project,
            features: features,
        },
    )
}

async function getFeatures({
    featureBoardProjectName,
    auth,
    interactive,
    apiEndpoint = 'https://api.featureboard.app',
}: {
    interactive: boolean
    auth: FeatureBoardAuth
    featureBoardProjectName?: string
    apiEndpoint?: string
}): Promise<[null | FeatureBoardProject, null | FeatureBoardFeature[]]> {
    const projectResults = await getProjects(apiEndpoint, auth)

    let project = projectResults.projects.find(
        (x: any) =>
            x.name.toLowerCase() ===
            featureBoardProjectName?.toLocaleLowerCase(),
    )

    if (!project && interactive) {
        if (projectResults.projects.length === 1) {
            project = projectResults.projects[0]
            console.log(
                `One project found. Setting FeatureBoard project to ${project.name}`,
            )
        } else {
            const promptResult = await prompts({
                type: 'select',
                name: 'project',
                message: `Pick your FeatureBoard project?`,
                validate: (x) =>
                    projectResults.projects.map((x) => x.name).includes(x),
                choices: projectResults.projects.map((x) => ({
                    title: x.name,
                    value: x,
                })),
            })

            if (!promptResult) {
                return [null, null]
            }

            project = promptResult.project
        }
    }
    if (!project) throw new Error('Unable to locate Project')

    const featuresResult = await getProjectFeatures(apiEndpoint, project, auth)

    return [project, featuresResult.features]
}
