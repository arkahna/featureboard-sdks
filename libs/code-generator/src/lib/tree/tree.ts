/* Copied from https://github.com/nrwl/nx/blob/ff04be51885b3ab2f635de6c13f3b29dec1859ba/packages/nx/src/generators/tree.ts with permission from the MIT licence on the 8/8/2023) */
import {
    Mode,
    chmodSync,
    mkdirSync,
    readFileSync,
    readdirSync,
    statSync,
    unlinkSync,
    writeFileSync,
} from 'fs'
import { dirname, join, relative, sep } from 'path'

/**
 * Options to set when writing a file in the Virtual file system tree.
 */
export interface TreeWriteOptions {
    /**
     * Permission to be granted on the file, given as a string (e.g `755`) or octal integer (e.g `0o755`).
     * The logical OR operator can be used to separate multiple permissions.
     * See https://nodejs.org/api/fs.html#fs_file_modes
     */
    mode?: Mode
}

/**
 * Virtual file system tree.
 */
export interface Tree {
    /**
     * Root of the workspace. All paths are relative to this.
     */
    root: string

    /**
     * Read the contents of a file.
     * @param filePath A path to a file.
     */
    read(filePath: string): Buffer | null

    /**
     * Read the contents of a file as string.
     * @param filePath A path to a file.
     * @param encoding the encoding for the result
     */
    read(filePath: string, encoding: BufferEncoding): string | null

    /**
     * Update the contents of a file or create a new file.
     */
    write(
        filePath: string,
        content: Buffer | string,
        options?: TreeWriteOptions,
    ): void

    /**
     * Check if a file exists.
     */
    exists(filePath: string): boolean

    /**
     * Delete the file.
     */
    delete(filePath: string): void

    /**
     * Rename the file or the folder.
     */
    rename(from: string, to: string): void

    /**
     * Check if this is a file or not.
     */
    isFile(filePath: string): boolean

    /**
     * Returns the list of children of a folder.
     */
    children(dirPath: string): string[]

    /**
     * Returns the list of currently recorded changes.
     */
    listChanges(): FileChange[]
}

/**
 * Description of a file change in the Nx virtual file system/
 */
export interface FileChange {
    /**
     * Path relative to the workspace root
     */
    path: string

    /**
     * Type of change: 'CREATE' | 'DELETE' | 'UPDATE'
     */
    type: 'CREATE' | 'DELETE' | 'UPDATE'

    /**
     * The content of the file or null in case of delete.
     */
    content: Buffer | null
    /**
     * Options to set on the file being created or updated.
     */
    options?: TreeWriteOptions
}

export class FsTree implements Tree {
    private recordedChanges: {
        [path: string]: {
            content: Buffer | null
            isDeleted: boolean
            options?: TreeWriteOptions
        }
    } = {}

    /**
     * Signifies if operations on the tree instance
     * are allowed. Set to false after changes are written
     * to disk, to prevent someone trying to use the tree to update
     * files when the tree is no longer effective.
     */
    private locked = false

    constructor(
        readonly root: string,
        private readonly isVerbose: boolean,
        private readonly logOperationId?: string,
    ) {}

    read(filePath: string): Buffer | null
    read(filePath: string, encoding: BufferEncoding): string | null
    read(filePath: string, encoding?: BufferEncoding): Buffer | string | null {
        filePath = this.normalize(filePath)
        try {
            let content: Buffer
            if (this.recordedChanges[this.rp(filePath)]) {
                content =
                    this.recordedChanges[this.rp(filePath)].content ??
                    Buffer.from('')
            } else {
                content = this.fsReadFile(filePath)
            }

            return encoding ? content.toString(encoding) : content
        } catch (e) {
            if (this.isVerbose) {
                console.error(e)
            }
            return null
        }
    }

    write(
        filePath: string,
        content: Buffer | string,
        options?: TreeWriteOptions,
    ): void {
        this.assertUnlocked()
        filePath = this.normalize(filePath)

        // Remove any recorded changes where a parent directory has been
        // deleted when writing a new file within the directory.
        let parent = dirname(this.rp(filePath))
        while (parent !== '.') {
            if (this.recordedChanges[parent]?.isDeleted) {
                delete this.recordedChanges[parent]
            }
            parent = dirname(parent)
        }

        if (
            this.fsExists(this.rp(filePath)) &&
            Buffer.from(content).equals(this.fsReadFile(filePath))
        ) {
            // Remove recorded change because the file has been restored to it's original contents
            delete this.recordedChanges[this.rp(filePath)]
            return
        }

        try {
            this.recordedChanges[this.rp(filePath)] = {
                content: Buffer.from(content),
                isDeleted: false,
                options,
            }
        } catch (e) {
            if (this.isVerbose) {
                console.error(e)
            }
        }
    }

    overwrite(
        filePath: string,
        content: Buffer | string,
        options?: TreeWriteOptions,
    ): void {
        filePath = this.normalize(filePath)
        this.write(filePath, content, options)
    }

    delete(filePath: string): void {
        this.assertUnlocked()
        filePath = this.normalize(filePath)
        if (this.filesForDir(this.rp(filePath)).length > 0) {
            this.filesForDir(this.rp(filePath)).forEach(
                (f) =>
                    (this.recordedChanges[f] = {
                        content: null,
                        isDeleted: true,
                    }),
            )
        }
        this.recordedChanges[this.rp(filePath)] = {
            content: null,
            isDeleted: true,
        }

        // Delete directory when is not root and there are no children
        if (
            filePath !== '' &&
            this.children(dirname(this.rp(filePath))).length < 1
        ) {
            this.delete(dirname(this.rp(filePath)))
        }
    }

    exists(filePath: string): boolean {
        filePath = this.normalize(filePath)
        try {
            if (this.recordedChanges[this.rp(filePath)]) {
                return !this.recordedChanges[this.rp(filePath)].isDeleted
            } else if (this.filesForDir(this.rp(filePath)).length > 0) {
                return true
            } else {
                return this.fsExists(filePath)
            }
        } catch {
            return false
        }
    }

    rename(from: string, to: string): void {
        this.assertUnlocked()
        from = this.normalize(from)
        to = this.normalize(to)
        if (from === to) {
            return
        }

        if (this.isFile(from)) {
            const content = this.read(this.rp(from))
            this.write(this.rp(to), content ?? Buffer.from(''))
            this.delete(this.rp(from))
        } else {
            for (const child of this.children(from)) {
                this.rename(join(from, child), join(to, child))
            }
        }
    }

    isFile(filePath: string): boolean {
        filePath = this.normalize(filePath)
        try {
            if (this.recordedChanges[this.rp(filePath)]) {
                return !this.recordedChanges[this.rp(filePath)].isDeleted
            } else {
                return this.fsIsFile(filePath)
            }
        } catch {
            return false
        }
    }

    children(dirPath: string): string[] {
        dirPath = this.normalize(dirPath)
        let res = this.fsReadDir(dirPath)

        res = [...res, ...this.directChildrenOfDir(this.rp(dirPath))]
        res = res.filter((q) => {
            const r =
                this.recordedChanges[this.normalize(join(this.rp(dirPath), q))]
            return !r?.isDeleted
        })
        // Dedupe
        return Array.from(new Set(res))
    }

    listChanges(): FileChange[] {
        const res = [] as FileChange[]
        Object.keys(this.recordedChanges).forEach((f) => {
            if (this.recordedChanges[f].isDeleted) {
                if (this.fsExists(f)) {
                    res.push({ path: f, type: 'DELETE', content: null })
                }
            } else {
                if (this.fsExists(f)) {
                    res.push({
                        path: f,
                        type: 'UPDATE',
                        content: this.recordedChanges[f].content,
                        options: this.recordedChanges[f].options,
                    })
                } else {
                    res.push({
                        path: f,
                        type: 'CREATE',
                        content: this.recordedChanges[f].content,
                        options: this.recordedChanges[f].options,
                    })
                }
            }
        })
        return res
    }

    private assertUnlocked() {
        if (this.locked) {
            console.error({
                title: `File changes have already been written to disk. Further changes were attempted ${
                    this.logOperationId
                        ? ` while running ${this.logOperationId}.`
                        : '.'
                }`,
                bodyLines: [
                    'The file system can no longer be modified. This commonly happens when a generator attempts to make further changes in its callback, or an asynchronous operation is still running after the generator completes.',
                ],
            })
            throw new Error('Tree changed after commit to disk.')
        }
    }

    private normalize(path: string) {
        return relative(this.root, join(this.root, path)).split(sep).join('/')
    }

    private fsReadDir(dirPath: string) {
        try {
            return readdirSync(join(this.root, dirPath))
        } catch {
            return []
        }
    }

    private fsIsFile(filePath: string): boolean {
        const stat = statSync(join(this.root, filePath))
        return stat.isFile()
    }

    private fsReadFile(filePath: string): Buffer {
        return readFileSync(join(this.root, filePath))
    }

    private fsExists(filePath: string): boolean {
        try {
            const stat = statSync(join(this.root, filePath))
            return stat.isFile() || stat.isDirectory()
        } catch {
            return false
        }
    }

    private filesForDir(path: string): string[] {
        return Object.keys(this.recordedChanges).filter(
            (f) =>
                f.startsWith(`${path}/`) && !this.recordedChanges[f].isDeleted,
        )
    }

    private directChildrenOfDir(path: string): string[] {
        const res: Record<string, boolean> = {}
        if (path === '') {
            return Object.keys(this.recordedChanges).map(
                (file) => file.split('/')[0],
            )
        }
        Object.keys(this.recordedChanges).forEach((f) => {
            if (f.startsWith(`${path}/`)) {
                // Remove the current folder's path from the directory
                const file = f.substring(path.length + 1)
                // Split the path on segments, and take the first one
                const basePath = file.split('/')[0]
                // Mark it as a child of the current directory
                res[basePath] = true
            }
        })

        return Object.keys(res)
    }

    private rp(pp: string): string {
        return pp.startsWith('/') ? pp.substring(1) : pp
    }
}

export function flushChanges(root: string, fileChanges: FileChange[]): void {
    fileChanges.forEach((f) => {
        const fpath = join(root, f.path)
        if (f.type === 'CREATE') {
            mkdirSync(dirname(fpath), { recursive: true })
            writeFileSync(fpath, f.content ?? Buffer.from('', 'utf8'))
            if (f.options?.mode) chmodSync(fpath, f.options.mode)
        } else if (f.type === 'UPDATE') {
            writeFileSync(fpath, f.content ?? Buffer.from('', 'utf8'))
            if (f.options?.mode) chmodSync(fpath, f.options.mode)
        } else if (f.type === 'DELETE') {
            unlinkSync(fpath)
        }
    })
}

export function printChanges(fileChanges: FileChange[], indent = ''): void {
    fileChanges.forEach((f) => {
        if (f.type === 'CREATE') {
            console.log(`${indent}${console.log('CREATE')} ${f.path}`)
        } else if (f.type === 'UPDATE') {
            console.log(`${indent}${console.log('UPDATE')} ${f.path}`)
        } else if (f.type === 'DELETE') {
            console.log(`${indent}${console.log('DELETE')} ${f.path}`)
        }
    })
}
