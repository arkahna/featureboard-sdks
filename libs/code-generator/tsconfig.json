{
    "extends": "../../tsconfig.settings.json",
    "compilerOptions": {
        "outDir": "./src",
        "rootDir": "./src",
        "types": [
            "node",
            "vitest/globals",
            "vitest/importMeta",
            "vite/client",
            "vitest"
        ]
    },
    "include": ["src/**/*.ts"],
    "exclude": ["node_modules", "tsc-out", "dist", "**/src/*.d.ts", "**/*.js"]
}
