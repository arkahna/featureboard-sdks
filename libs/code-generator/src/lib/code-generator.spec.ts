import * as fs from 'fs/promises'
import { HttpResponse, http } from 'msw'
import type { SetupServer } from 'msw/node'
import { setupServer } from 'msw/node'
import * as path from 'path'
import type { CodeGeneratorOptions } from './code-generator'
import { codeGenerator } from './code-generator'
import type { Tree } from './tree/tree'
import { FsTree } from './tree/tree'

describe('code-generator', () => {
    let tree: Tree
    let options: Omit<CodeGeneratorOptions, 'tree' | 'relativeFilePath'> & {
        subFolder: string
    }
    let server: SetupServer

    beforeAll(() => {
        server = setupServer(
            http.get('https://api.featureboard.app/projects', async () => {
                const file = await fs.readFile(
                    path.join(__dirname, './test-data/projects.json'),
                    {
                        encoding: 'utf8',
                    },
                )
                return HttpResponse.json(JSON.parse(file))
            }),
            http.get(
                'https://api.featureboard.app/projects/ivu5ntxbxeeorm248jl25bs6/features',
                async () => {
                    const file = await fs.readFile(
                        path.join(
                            __dirname,
                            './test-data/features-ivu5ntxbxeeorm248jl25bs6.json',
                        ),
                        {
                            encoding: 'utf8',
                        },
                    )
                    return HttpResponse.json(JSON.parse(file))
                },
            ),
        )
        server.listen({ onUnhandledRequest: 'error' })
    })

    beforeEach(async () => {
        const cwd = process.cwd()
        tree = new FsTree(cwd, true)

        const subFolder = 'apps/my-app'
        tree.write(
            path.join(subFolder, 'Project.FeatureBoard.MyApp.csproj'),
            '',
        )

        options = {
            subFolder: subFolder,
            template: 'dotnet-api',
            interactive: false,
            featureBoardProjectName: 'SaaSy Icons',
            auth: { featureBoardApiKey: 'This is totally a key' },
        }
    })

    describe('template: dotnet-api', () => {
        beforeEach(async () => {
            options.template = 'dotnet-api'
        })

        it('should run successfully', async () => {
            await codeGenerator({
                tree: tree,
                relativeFilePath: options.subFolder,
                ...options,
            })
        })

        it('should produce the expected features files', async () => {
            await codeGenerator({
                tree: tree,
                relativeFilePath: options.subFolder,
                ...options,
            })

            const files = tree
                .listChanges()
                .filter((x) => x.path.match(/^apps\/my-app\/.*$/))

            expect(files).toHaveLength(2)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })

    describe('template: typescript', () => {
        beforeEach(async () => {
            options.template = 'typescript'
        })

        it('should run successfully', async () => {
            await codeGenerator({
                tree: tree,
                relativeFilePath: options.subFolder,
                ...options,
            })
        })

        it('should produce the expected features files', async () => {
            await codeGenerator({
                tree: tree,
                relativeFilePath: options.subFolder,
                ...options,
            })

            const files = tree
                .listChanges()
                .filter((x) => x.path.match(/^apps\/my-app\/.*$/))

            expect(files).toHaveLength(2)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })

    afterEach(() => {
        server.resetHandlers()
    })

    afterAll(() => {
        server.close()
    })
})
