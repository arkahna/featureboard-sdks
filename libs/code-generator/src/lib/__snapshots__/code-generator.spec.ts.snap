// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`code-generator > template: dotnet-api > should produce the expected features files > apps/my-app/Features.cs 1`] = `
"using FeatureBoard.DotnetSdk.Attributes;
using FeatureBoard.DotnetSdk.Models;
using System.Runtime.Serialization;

namespace Project.FeatureBoard.MyApp;

// Features
public class Features : IFeatures
{
  [FeatureKeyName(\\"requires-attribution\\")]
  public bool RequiresAttribution { get; set; }
  [FeatureKeyName(\\"allow-edits\\")]
  public bool AllowEdits { get; set; }
  [FeatureKeyName(\\"number-custom-collections\\")]
  public decimal NumberCustomCollections { get; set; }
  [FeatureKeyName(\\"team-management\\")]
  public bool TeamManagement { get; set; }
  [FeatureKeyName(\\"custom-collections\\")]
  public CustomCollections CustomCollections { get; set; }
  [FeatureKeyName(\\"view-subscribers\\")]
  public bool ViewSubscribers { get; set; }
}


// Options Enums

//Custom collections
public enum CustomCollections
{
  [EnumMember(Value = \\"Disabled\\")]
  Disabled,
  [EnumMember(Value = \\"Enabled\\")]
  Enabled,
  [EnumMember(Value = \\"Shared\\")]
  Shared,
}

// Type Enums
public enum BooleanFeature
{
  RequiresAttribution,
  AllowEdits,
  TeamManagement,
  ViewSubscribers,
}
public enum NumberFeature
{
  NumberCustomCollections,
}
public enum StringFeature
{
}

// Feature Gate attribute
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class FeatureFilterAttribute : FeatureFilterAttributeBase
{
  protected override string Feature { get; }
  protected override object DefaultValue { get; }
  protected override object? IsEqualTo { get; }
  private readonly Dictionary<string, string> _featureKeyLookup = new()
  {
    { \\"RequiresAttribution\\", \\"requires-attribution\\" },
    { \\"AllowEdits\\", \\"allow-edits\\" },
    { \\"NumberCustomCollections\\", \\"number-custom-collections\\" },
    { \\"TeamManagement\\", \\"team-management\\" },
    { \\"ViewSubscribers\\", \\"view-subscribers\\" },
  };

  public FeatureFilterAttribute(BooleanFeature feature, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(BooleanFeature feature, bool isEqualTo, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(StringFeature feature, string isEqualTo, string defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(NumberFeature feature, decimal isEqualTo, decimal defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(CustomCollections isEqualTo, CustomCollections defaultValue)
  {
    Feature = \\"custom-collections\\";
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

}
"
`;

exports[`code-generator > template: dotnet-api > should produce the expected features files > apps/my-app/Project.FeatureBoard.MyApp.csproj 1`] = `""`;

exports[`code-generator > templateType: dotnet-api > should produce the expected features files > apps/my-app/Features.cs 1`] = `
"using FeatureBoard.DotnetSdk.Attributes;
using FeatureBoard.DotnetSdk.Models;
using System.Runtime.Serialization;

namespace Project.FeatureBoard.MyApp;

// Features
public class Features : IFeatures
{
  [FeatureKeyName(\\"requires-attribution\\")]
  public bool RequiresAttribution { get; set; }
  [FeatureKeyName(\\"allow-edits\\")]
  public bool AllowEdits { get; set; }
  [FeatureKeyName(\\"number-custom-collections\\")]
  public decimal NumberCustomCollections { get; set; }
  [FeatureKeyName(\\"team-management\\")]
  public bool TeamManagement { get; set; }
  [FeatureKeyName(\\"custom-collections\\")]
  public CustomCollections CustomCollections { get; set; }
  [FeatureKeyName(\\"view-subscribers\\")]
  public bool ViewSubscribers { get; set; }
}


// Options Enums

//Custom collections
public enum CustomCollections
{
  [EnumMember(Value = \\"Disabled\\")]
  Disabled,
  [EnumMember(Value = \\"Enabled\\")]
  Enabled,
  [EnumMember(Value = \\"Shared\\")]
  Shared,
}

// Type Enums
public enum BooleanFeature
{
  RequiresAttribution,
  AllowEdits,
  TeamManagement,
  ViewSubscribers,
}
public enum NumberFeature
{
  NumberCustomCollections,
}
public enum StringFeature
{
}

// Feature Gate attribute
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class FeatureFilterAttribute : FeatureFilterAttributeBase
{
  protected override string Feature { get; }
  protected override object DefaultValue { get; }
  protected override object? IsEqualTo { get; }
  private readonly Dictionary<string, string> _featureKeyLookup = new()
  {
    { \\"RequiresAttribution\\", \\"requires-attribution\\" },
    { \\"AllowEdits\\", \\"allow-edits\\" },
    { \\"NumberCustomCollections\\", \\"number-custom-collections\\" },
    { \\"TeamManagement\\", \\"team-management\\" },
    { \\"ViewSubscribers\\", \\"view-subscribers\\" },
  };

  public FeatureFilterAttribute(BooleanFeature feature, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(BooleanFeature feature, bool isEqualTo, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(StringFeature feature, string isEqualTo, string defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(NumberFeature feature, decimal isEqualTo, decimal defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(CustomCollections isEqualTo, CustomCollections defaultValue)
  {
    Feature = \\"custom-collections\\";
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

}
"
`;

exports[`code-generator > templateType: dotnet-api > should produce the expected features files > apps/my-app/Project.FeatureBoard.MyApp.csproj 1`] = `""`;

exports[`code-generator templateType: dotnet-api should produce the expected features files: apps/my-app/Features.cs 1`] = `
"using FeatureBoard.DotnetSdk.Attributes;
using FeatureBoard.DotnetSdk.Models;
using System.Runtime.Serialization;

namespace Project.FeatureBoard.MyApp;

// Features
public class Features : IFeatures
{
  [FeatureKeyName("requires-attribution")]
  public bool RequiresAttribution { get; set; }
  [FeatureKeyName("allow-edits")]
  public bool AllowEdits { get; set; }
  [FeatureKeyName("number-custom-collections")]
  public decimal NumberCustomCollections { get; set; }
  [FeatureKeyName("team-management")]
  public bool TeamManagement { get; set; }
  [FeatureKeyName("custom-collections")]
  public CustomCollections CustomCollections { get; set; }
  [FeatureKeyName("view-subscribers")]
  public bool ViewSubscribers { get; set; }
}


// Options Enums

//Custom collections
public enum CustomCollections
{
  [EnumMember(Value = "Disabled")]
  Disabled,
  [EnumMember(Value = "Enabled")]
  Enabled,
  [EnumMember(Value = "Shared")]
  Shared,
}

// Type Enums
public enum BooleanFeature
{
  RequiresAttribution,
  AllowEdits,
  TeamManagement,
  ViewSubscribers,
}
public enum NumberFeature
{
  NumberCustomCollections,
}
public enum StringFeature
{
}

// Feature Gate attribute
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class FeatureFilterAttribute : FeatureFilterAttributeBase
{
  protected override string Feature { get; }
  protected override object DefaultValue { get; }
  protected override object? IsEqualTo { get; }
  private readonly Dictionary<string, string> _featureKeyLookup = new()
  {
    { "RequiresAttribution", "requires-attribution" },
    { "AllowEdits", "allow-edits" },
    { "NumberCustomCollections", "number-custom-collections" },
    { "TeamManagement", "team-management" },
    { "ViewSubscribers", "view-subscribers" },
  };

  public FeatureFilterAttribute(BooleanFeature feature, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(BooleanFeature feature, bool isEqualTo, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(StringFeature feature, string isEqualTo, string defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(NumberFeature feature, decimal isEqualTo, decimal defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(CustomCollections isEqualTo, CustomCollections defaultValue)
  {
    Feature = "custom-collections";
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

}
"
`;

exports[`code-generator templateType: dotnet-api should produce the expected features files: apps/my-app/Project.FeatureBoard.MyApp.csproj 1`] = `""`;
